README.md

```diff
@@ -669,6 +669,105 @@ you've fixed the error, it's usually safe to re-run the `spin-up` or `tear-down`
+```
+
+When Terragrunt finds a `terraform` block, it will:
+ 
+1. Download the configurations specified via the `source` parameter into a temporary folder. This is done using the
+   [terraform init command](https://www.terraform.io/docs/commands/init.html), so the `source` parameter supports the
+   exact same syntax as the [module source](https://www.terraform.io/docs/modules/sources.html) parameter, including
+   local file paths, Git URLs, and Git URLs with `ref` parameters (useful for checking out a specific tag, commit, or
+   branch of Git repo).
+1. Copy all files from the current working directory into the temporary folder. This is useful if you have `.tfvars`
+   files in your local path that you are using to set variables within your templates.
+1. Execute whatever Terraform command you specified in that temporary folder. **Note**: if you are passing any file
+   paths (other than paths to files in the current working directory) to Terraform via command-line options, those 
+   paths must be absolute paths since we will be running Terraform from the temporary folder!
+
+With this approach, you could define all of your Terraform code in one repo called, for example, 
+infrastructure-modules:
+
+```
+infrastructure-modules
+  └ frontend-app
+      └ main.tf
+      └ vars.tf
+      └ outputs.tf
+  └ backend-app
+  └ search-app
+  └ mysql
+  └ redis
+  └ vpc
+```
+
+You could then use this infrastructure in a separate repo called, for example, infrastructure-live, as follows:
+
+```
+infrastructure-live
+  └ stage
+    └ frontend-app
+        └ .terragrunt
+        └ terraform.tfvars
+    └ backend-app
+    └ search-app
+    └ mysql
+    └ redis
+    └ vpc
+  └ prod
+    └ frontend-app
+        └ .terragrunt
+        └ terraform.tfvars
+    └ backend-app
+    └ search-app
+    └ mysql
+    └ redis
+    └ vpc
+```
+
+Each `.terragrunt` file would set the `source` parameter to a specific version of the relevant module in 
+infrastructure-modules:
+
+```hcl
+terraform {
+  source = "git::git@github.com:my-company/infrastructure-modules.git//frontend-app?ref=v0.0.1"
+}
+```
+
+The `terraform.tfvars` files would provide all the variables needed for that module:
+
+```hcl
+app_name = "my-frontend-app"
+num_instances = 5
+instance_type = "m4.large"
+```
+
+Now, all of your infrastructure is versioned, and there is no duplication of Terraform code to use it across many
+environments. You just run `terragrunt apply` and Terragrunt will check out the specified version of the code and run
+`terraform apply` on it.
+
+Note that you can also use the `--terragrunt-source` command-line option and `TERRAGRUNT_SOURCE` environment variable
+to override the `source` parameter. This is useful to point Terragrunt at a local checkout of your code so you can do 
+rapid, iterative, make-a-change-and-rerun development:
+   
+```
+cd infrastructure-live/stage/frontend-app
+terragrunt apply --terragrunt-source ../../../infrastructure-modules/frontend-app
+```   
+
```

Comment:
First, this is fantastic. I am really excited to start playing around with this.

Second, just a sanity check. So, let's say one of our customers is using our vpc modules. Then here's what the chain of calls looks like:

- client/infrastructure-live: `vpc`
- client/infrastructure-modules: `vpc`
- gruntwork/module-vpc: `vpc-app`

So to state the obvious, this adds one more layer to our module calls than we had before. But I don't really see any downsides to this. Of course, this does raise the question of how we handle our traditional infrastructure-modules, like `ecs-cluster`. For example:

- client/infrastructure-live: `stage/ecs-cluster`
- client/infrastructure-modules: `ecs-cluster`
- client/infrastructure-modules: `modules/ecs-cluster`
- gruntwork/module-ecs: `ecs-cluster`

It means one infrastructure-module might want to make a versioned reference to another infrastructure-module. How are you addressing that? Might be better for offline or out-of-band discussion since it doesn't directly pertain to this PR.